x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authc.saml;
+
+import org.elasticsearch.common.unit.TimeValue;
+import org.opensaml.saml.saml2.core.LogoutResponse;
+import org.w3c.dom.Element;
+
+import java.time.Clock;
+import java.util.Collection;
+
+import static org.elasticsearch.xpack.security.authc.saml.SamlUtils.samlException;
+
+public class SamlLogoutResponseHandler extends SamlResponseHandler {
+
+    private static final String LOGOUT_RESPONSE_TAG_NAME = "LogoutResponse";
+
+    public SamlLogoutResponseHandler(
+        Clock clock, IdpConfiguration idp, SpConfiguration sp, TimeValue maxSkew) {
+        super(clock, idp, sp, maxSkew);
+    }
+
+    public void handle(boolean httpRedirect, String payload, Collection<String> allowedSamlRequestIds) {
+        final ParsedQueryString parsed = parseQueryStringAndValidateSignature(payload, "SAMLResponse");
+        if (httpRedirect && parsed.hasSignature == false) {
+            throw samlException("URL is not signed, but is required for HTTP-Redirect binding");
+        } else if (httpRedirect == false && parsed.hasSignature) {
+            throw samlException("URL is signed, but binding is HTTP-POST");
+        }
+
+        final Element root;
+        if (httpRedirect) {
+            logger.info("Process SAML LogoutResponse with HTTP-Redirect binding");
+            root = parseSamlMessage(inflate(decodeBase64(parsed.samlMessage)));
+        } else {
+            logger.info("Process SAML LogoutResponse with HTTP-POST binding");
+            root = parseSamlMessage(decodeBase64(parsed.samlMessage));
+        }
+
+        if (LOGOUT_RESPONSE_TAG_NAME.equals(root.getLocalName()) && SAML_NAMESPACE.equals(root.getNamespaceURI())) {
+            final LogoutResponse logoutResponse = buildXmlObject(root, LogoutResponse.class);
+            if (logoutResponse == null) {
```

Comment:
Indeed `buildXmlObject` does not return `null`. I'll remove it. Just for discussion: Since Java does not have proper null safty, aggressively checking for `null` return value could be argued as defensive coding or invariant enforcing. I think this could be a reason that we have many `assert xxx` statement throughout the code base. The difference is that they do not run in production but helps to catch potential code change gaps in testing. What would be your take on this?

Reply:
We do have asserts throughout the code base and we do have places where we add extra null checks "just to be safe", but we have been inconsistent in doing so I think. I love defensive coding as the next person but I don't know where the line should be drawn so that we don't have a `if (x == null)` check every 5 lines. Thinking out loud, we should be able to trust the methods we write to not return null unless they explicitly document that and handle null checks in the method instead of everywhere we call that method

Tim had shared some thoughts on `assert`s here: https://github.com/elastic/elasticsearch/pull/52936#discussion_r386167255

Reply:
Thanks. The difference is subtle but insightful.